File: Diagrammatize.Design.Pattern/code/python/Command/A2/MainWindow.py

import tkinter as tk
from command import MacroCommand
from drawer import DrawCommand, DrawCanvas

class MainWindow(tk.Frame):
    def __init__(self, parent):
        tk.Frame.__init__(self, parent)

        buttonBox = tk.Frame(self)
        self.clearButton = tk.Button(buttonBox, text="clear", command=self.onClearButtonClicked)
        self.clearButton.pack(side=tk.LEFT, fill=tk.X, expand=1)

        self.undoButton = tk.Button(buttonBox, text="undo", command=self.onUndoButtonClicked)
        self.undoButton.pack(side=tk.LEFT, fill=tk.X, expand=1)
        buttonBox.pack(fill=tk.X)

        self.history = MacroCommand()
        self.canvas = DrawCanvas(self, 400, 400, self.history)
        self.canvas.pack()

        self.canvas.bind("<B1-Motion>", self.mouseDragged)

    def onClearButtonClicked(self):
        self.history.clear()
        self.canvas.repaint()

    def onUndoButtonClicked(self):
        self.history.undo()
        self.canvas.repaint()

    def mouseDragged(self, event):
        cmd = DrawCommand(self.canvas, event)
        self.history.append(cmd)
        cmd.execute()

File: Diagrammatize.Design.Patterns/code/cxx/Command/A2/MainWindow.cpp
#include "MainWindow.hpp"
#include <QHBoxLayout>
#include <QVBoxLayout>
#include <QEvent>
#include <QMouseEvent>
#include "drawer/DrawCommand.hpp"

MainWindow::MainWindow(const QString& title, QWidget *parent)
	: QWidget(parent)
{
    history = std::shared_ptr<MacroCommand>(new MacroCommand);

    setWindowTitle(title);    

    QHBoxLayout *buttonBox = new QHBoxLayout;

    clearButton = new QPushButton("clear", this);
    buttonBox->addWidget(clearButton);

    undoButton = new QPushButton("undo", this);
    buttonBox->addWidget(undoButton);

    QVBoxLayout *mainBox = new QVBoxLayout();

    mainBox->addLayout(buttonBox);
    
    canvas = new DrawCanvas(400, 400, history);
    mainBox->addWidget(canvas);

    setLayout(mainBox);

//    setMinimumSize(600,400);

    connect(clearButton, SIGNAL(clicked()), this, SLOT(onClearButtonClicked()));
    connect(undoButton, SIGNAL(clicked()), this, SLOT(onUndoButtonClicked()));
	canvas->installEventFilter(this);
}

bool MainWindow::eventFilter(QObject* watched,QEvent* event)
{
	if(watched == canvas)
	{
		if(event->type() == QEvent::MouseMove)
		{
			QMouseEvent *mouseEvent = (QMouseEvent *)event;	
            mouseDragged(mouseEvent);
		}		
	}

	return QWidget::eventFilter(watched,event);	
}

void MainWindow::mouseDragged(QMouseEvent *e)
{
    std::shared_ptr<Command> cmd(new DrawCommand(canvas, e->pos()));
    history->append(cmd);
    cmd->execute();
}

void MainWindow::onClearButtonClicked()
{
    history->clear();
    canvas->repaint();
}

void MainWindow::onUndoButtonClicked()
{
    history->undo();
    canvas->repaint();
}

